File: backend/app/api/readiness.py
from uuid import UUID

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.auth import get_current_user
from app.database import get_db
from app.models.user import User
from app.services import readiness_service

router = APIRouter(tags=["readiness"])


class CheckResultResponse(BaseModel):
    name: str
    status: str
    message: str
    count: int | None = None
    required: int | None = None


class ReadinessResponse(BaseModel):
    ready: bool
    checks: list[CheckResultResponse]


class ExamSummaryResponse(BaseModel):
    roster_count: int
    assigned_count: int
    unassigned_count: int
    room_count: int
    generated_count: int
    qr_count: int
    has_template: bool
    template_version: int | None


@router.get("/api/exams/{exam_id}/readiness", response_model=ReadinessResponse)
def get_readiness(
    exam_id: UUID,
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    result = readiness_service.validate_exam_readiness(db, exam_id)
    return ReadinessResponse(
        ready=result.ready,
        checks=[
            CheckResultResponse(
                name=c.name,
                status=c.status,
                message=c.message,
                count=c.count,
                required=c.required,
            )
            for c in result.checks
        ],
    )


@router.get("/api/exams/{exam_id}/summary", response_model=ExamSummaryResponse)
def get_exam_summary(
    exam_id: UUID,
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    return readiness_service.get_exam_summary(db, exam_id)

File: frontend/src/lib/api/readiness.ts
const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export interface CheckResult {
  name: string;
  status: "PASS" | "FAIL" | "WARN";
  message: string;
  count: number | null;
  required: number | null;
}

export interface ReadinessResult {
  ready: boolean;
  checks: CheckResult[];
}

export interface ExamSummary {
  roster_count: number;
  assigned_count: number;
  unassigned_count: number;
  room_count: number;
  generated_count: number;
  qr_count: number;
  has_template: boolean;
  template_version: number | null;
}

export async function getReadiness(examId: string): Promise<ReadinessResult> {
  const response = await fetch(`${API_BASE}/api/exams/${examId}/readiness`, {
    credentials: "include",
  });
  if (!response.ok) throw new Error("Failed to check readiness");
  return response.json();
}

export async function getExamSummary(examId: string): Promise<ExamSummary> {
  const response = await fetch(`${API_BASE}/api/exams/${examId}/summary`, {
    credentials: "include",
  });
  if (!response.ok) throw new Error("Failed to get summary");
  return response.json();
}
